R/workflows.R

```diff
@@ -106,38 +110,59 @@ construct_workflow <- function(reports, report_names, dependencies) {
       report_detail
     })
   }
+
   workflow <- lapply(order, build_item)
   unlist(workflow, recursive = FALSE, use.names = FALSE)
 }
 
-## This algorithm comes from here:
-## http://blog.jupo.org/2012/04/06/topological-sorting-acyclic-directed-graphs/
-## and assumes that the graph is expressed as a *named* list.  The
-## daughters of an element are its dependencies.
 topological_sort <- function(graph) {
-  m <- matrix(FALSE, length(graph), length(graph))
-  for (i in seq_along(graph)) {
-    m[, i] <- unname(names(graph) %in% graph[[i]])
-  }
-  pending <- rep(TRUE, length(graph))
-  graph_sorted <- integer(0)
-  while (any(pending)) {
-    i <- which(pending)[colSums(m[, pending, drop = FALSE]) == 0]
-    if (length(i) > 0L) {
-      graph_sorted <- c(graph_sorted, i)
-      pending[i] <- FALSE
-      m[i, ] <- FALSE
-    } else {
+  report_names <- names(graph)
+  graph_sorted <- NULL
+  explored <- rep(FALSE, length(graph))
+  deps <- lapply(seq_along(graph),
+                 function(i) which(report_names %in% graph[[i]]))
+
+  stack <- 1
+
+  while (any(!explored)) {
+    if (length(stack) > 1 && head(stack, n = 1) == tail(stack, n = 1)) {
       f <- function(i) {
         sprintf("  %s: depends on %s",
-                names(graph)[[i]], paste(err[m[pending, i]], collapse = ", "))
+                names(graph)[[i]], paste(graph[[i]], collapse = ", "))
+      }
+
+      stack <- rev(stack)
+      node <- stack[1]
+      nodes <- node
+      stack <- stack[-1]
+      while (length(stack) > 1) {
+        children <- which(report_names %in% graph[[node]])
+        index <- seq_along(children)
+        stack <- stack[-index]
+        node <- head(children, n = 1)
```

Comment:
Is the use of `head` here to gracefully return nothing when there are no children?

Reply:
Ha, no, I don't know why I used it